src/net_processing.cpp

```diff
@@ -1294,19 +1302,32 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
     });
 }
 
-bool PeerManagerImpl::TipMayBeStale()
+PeerManagerImpl::Staleness PeerManagerImpl::TipMayBeStale()
 {
     AssertLockHeld(cs_main);
     const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
     if (m_last_tip_update.load() == 0s) {
         m_last_tip_update = GetTime<std::chrono::seconds>();
     }
-    return m_last_tip_update.load() < GetTime<std::chrono::seconds>() - std::chrono::seconds{consensusParams.nPowTargetSpacing * 3} && mapBlocksInFlight.empty();
+
+    auto last_update_time = m_last_tip_update.load();
+    if (last_update_time < GetTime<std::chrono::seconds>() - std::chrono::seconds{consensusParams.nPowTargetSpacing * 3} && mapBlocksInFlight.empty()) {
+        // If tip is further than NODE_NETWORK_LIMITED_MIN_BLOCKS, we are really behind and shouldn't connect to limited peers anymore.
+        Assume(NODE_NETWORK_LIMITED_MIN_BLOCKS > 3);
+        bool in_ibd = NodeSeconds{last_update_time} < GetAdjustedTime() - std::chrono::seconds{consensusParams.nPowTargetSpacing * NODE_NETWORK_LIMITED_MIN_BLOCKS};
```

Comment:
636d927c4a9cd5cbf4d0c22e1c25dc44ecaf7a69

Is it still fair to call this state *initial* block download?

Reply:
That's just me trying to save few characters. Will rename it to `past_net_limited`.

Reply:
> Any reason to use adjusted time, when the tip update time is recorded as NodeClock?

Good point. The idea was to considerate the peers' time too, so the node has less chances of requesting a block at the limited peers threshold that does no longer exist in the other side. In other words, if the peer's time deviates slightly from the node time, the peer could have pruned the block at the verge of the threshold.
But we could achieve the same outcome by reducing the window size by one or two blocks. Without requiring to mix `GetTime` and `GetAdjustedTime` usages.

Reply:
> > In other words, if the peer's time deviates slightly from the node time
> 
> I don't think `GetAdjustedTime` offers this precision even. Also, it is not a per-peer offset, but a global one.

Yeah I know. I just thought it was better to do that than nothing.
Still, I'm quite sure we cannot trigger the edge case I mentioned above only using core vanilla alone. The pruning process isn't that aggressive for the time being. It might arise across different implementations but we can revisit it in a follow-up, the limited peers connection window can be easily shortened without any problem.